sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/joins.scala

```diff
@@ -152,3 +152,71 @@ object EliminateOuterJoin extends Rule[LogicalPlan] with PredicateHelper {
       if (j.joinType == newJoinType) f else Filter(condition, j.copy(joinType = newJoinType))
   }
 }
+
+/**
+ * A rule that uses propagated constraints to infer join conditions. The optimization is applicable
+ * only to CROSS joins.
+ *
+ * For instance, if there is a CROSS join, where the left relation has 'a = 1' and the right
+ * relation has 'b = 1', then the rule infers 'a = b' as a join predicate.
+ */
+object InferJoinConditionsFromConstraints extends Rule[LogicalPlan] with PredicateHelper {
+
+  def apply(plan: LogicalPlan): LogicalPlan = {
+    if (SQLConf.get.constraintPropagationEnabled) {
+      inferJoinConditions(plan)
+    } else {
+      plan
+    }
+  }
+
+  private def inferJoinConditions(plan: LogicalPlan): LogicalPlan = plan transform {
+    case join @ Join(left, right, Cross, conditionOpt) =>
+      val leftConstraints = join.constraints.filter(_.references.subsetOf(left.outputSet))
+      val rightConstraints = join.constraints.filter(_.references.subsetOf(right.outputSet))
```

Comment:
I don't think we need to separate the constraints as left only and right only.
The following case can infer `t1.col1 = t2.col1`:
```scala
Seq((1, 2)).toDF("col1", "col2").write.saveAsTable("t1")
Seq((1, 2)).toDF("col1", "col2").write.saveAsTable("t2")
val df = spark.sql("SELECT * FROM t1 CROSS JOIN t2 ON t1.col1 >= t2.col1 " +
   "WHERE t1.col1 = t1.col2 + t2.col2 and t2.col1 = t1.col2 + t2.col2")
```

Reply:
@gengliangwang Yeah, makes sense. So, ``PushPredicateThroughJoin`` would push the where clause into the join and the proposed rule will infer ``t1.col1 = t2.col1`` and change the join type to INNER. As a result, the final join condition will be ``t1.col1 = t2.col1 and t1.col1 >= t2.col1 and (t1.col1 = t1.col2 + t2.col2 and t2.col1 = t1.col2 + t2.col2)``. Am I right?

Reply:
Yes, you are right.